### PR TITLE
[AMDGPU][RFC] Combine asm and disasm tests.

### PR DESCRIPTION
**Pull request by kosarev**
Eliminates the need to replicate the same instructions in MC and MC/Disassembler tests and synchronize changes in them. Also highlights differences between disassembled, reassembled and original instructions.

### CODE REVIEW

**Comment by MaskRay** [comment]
I hope that we do not add new MCStreamer derivatives just for testing purposes.
For round tripping tests, can you do `llvm-mc ... -filetype=obj | llvm-objdump -d -` instead?

**Review by MaskRay — CHANGES_REQUESTED** [review]
.

**Comment by kosarev** [comment]
@MaskRay `llvm-objdump -d` doesn't use the asm streamer, meaning it would require separate check lines, whereas the idea is to make same check lines work for both the assembler and disassembler tests.

Regarding round-tripping, we may want to consider that at some point, but we still need to directly check what the disassembler produces, because reassembling may mask problems.

**Comment by kosarev** [comment]
> Can you give an example that llvm-objdump is insufficient for your case?

Using `llvm-objdump` would require having two separate check lines for every instruction, avoiding which is the point of the patch.

```
v_bfrev_b32_e32 v5, v1
// GFX12-ASM: v_bfrev_b32_e32 v5, v1                  ; encoding: [0x01,0x71,0x0a,0x7e]
// GFX12-DIS: v_bfrev_b32_e32 v5, v1                                     // 000000000000: 7E0A7101
```

**Comment by MaskRay** [comment]
> > Can you give an example that llvm-objdump is insufficient for your case?
> 
> Using `llvm-objdump` would require having two separate check lines for every instruction, avoiding which is the point of the patch.
> 
> ```
> v_bfrev_b32_e32 v5, v1
> // GFX12-ASM: v_bfrev_b32_e32 v5, v1                  ; encoding: [0x01,0x71,0x0a,0x7e]
> // GFX12-DIS: v_bfrev_b32_e32 v5, v1                                     // 000000000000: 7E0A7101
> ```

Utilize `-SAME:` directives?
```
# RUN: llvm-mc ... %s | FileCheck
# RUN: llvm-mc -filetype=obj ... %s | llvm-objdump -d - | FileCheck --check-prefixes=CHECK,ENCODING

# CHECK: nop
# ENCODING-SAME: [0,0,0,0]
nop
```

**Comment by kosarev** [comment]
> Utilize `-SAME:` directives?
> 
> ```
> # RUN: llvm-mc ... %s | FileCheck
> # RUN: llvm-mc -filetype=obj ... %s | llvm-objdump -d - | FileCheck --check-prefixes=CHECK,ENCODING
> 
> # CHECK: nop
> # ENCODING-SAME: [0,0,0,0]
> nop
> ```

Again, this wouldn't work because `llvm-objdump -d` does not print encodings in the `[0,0,0,0]` form.

**Comment by kosarev** [comment]
Also, the first RUN line doesn't check the resulting instruction code, so doesn't work as an assembler-side test.

**Comment by MaskRay** [comment]
> > Utilize `-SAME:` directives?
> > ```
> > # RUN: llvm-mc ... %s | FileCheck
> > # RUN: llvm-mc -filetype=obj ... %s | llvm-objdump -d - | FileCheck --check-prefixes=CHECK,ENCODING
> > 
> > # CHECK: nop
> > # ENCODING-SAME: [0,0,0,0]
> > nop
> > ```
> 
> Again, this wouldn't work because `llvm-objdump -d` does not print encodings in the `[0,0,0,0]` form.

llvm-objdump seems to places the encodings as comments. You can check them with `-SAME:`.

> Also, the first RUN line doesn't check the resulting instruction code, so doesn't work as an assembler-side test.

I am confused. Why is `llvm-mc -filetype=obj ... %s | llvm-objdump -d - ` not checking the result?

File: llvm/lib/MC/MCHexStreamer.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+void MCHexStreamer::emitInstruction(const MCInst &Inst,
+                                    const MCSubtargetInfo &STI) {
+  SmallString<256> Code;
+  SmallVector<MCFixup, 4> Fixups;
+  getAssembler().getEmitter().encodeInstruction(Inst, Code, Fixups, STI);
+
+  OS << "[";
+  for (auto [I, B] : enumerate(Code))
+    OS << (I != 0 ? "," : "") << format("0x%02x", B & 0xff);
```
Comment (arsenm) [inline llvm/lib/MC/MCHexStreamer.cpp:71]: Use ListSeparator for the =0 case

File: llvm/lib/MC/MCHexStreamer.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+} // end anonymous namespace.
+
+void MCHexStreamer::emitInstruction(const MCInst &Inst,
+                                    const MCSubtargetInfo &STI) {
+  SmallString<256> Code;
+  SmallVector<MCFixup, 4> Fixups;
+  getAssembler().getEmitter().encodeInstruction(Inst, Code, Fixups, STI);
+
+  OS << "[";
```
Comment (arsenm) [inline llvm/lib/MC/MCHexStreamer.cpp:69]:
```suggestion
  OS << '[';
```

File: llvm/lib/MC/MCHexStreamer.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  for (auto [I, B] : enumerate(Code))
+    OS << (I != 0 ? "," : "") << format("0x%02x", B & 0xff);
+  OS << "]  #";
+
+  if (getTargetStreamer())
+    getTargetStreamer()->prettyPrintAsm(*InstPrinter, 0, Inst, STI, OS);
+  else
+    InstPrinter->printInst(&Inst, 0, "", STI, OS);
+
+  OS << "\n";
```
Comment (arsenm) [inline llvm/lib/MC/MCHexStreamer.cpp:79]:
```suggestion
  OS << '\n';
```

File: llvm/tools/llvm-mc/llvm-mc.cpp [hunk 4]
```diff
@@ -541,11 +543,13 @@ int main(int argc, char **argv) {
...
-        TheTarget->createAsmStreamer(Ctx, std::move(FOut), /*asmverbose*/ true,
-                                     /*useDwarfDirectory*/ true, IP,
-                                     std::move(CE), std::move(MAB), ShowInst));
-
+    Str.reset(FileType == OFT_HexFile
+                  ? TheTarget->createHexStreamer(Ctx, std::move(FOut), IP,
+                                                 std::move(CE), std::move(MAB))
+                  : TheTarget->createAsmStreamer(
+                        Ctx, std::move(FOut), /*asmverbose*/ true,
+                        /*useDwarfDirectory*/ true, IP, std::move(CE),
```
Comment (arsenm) [inline llvm/tools/llvm-mc/llvm-mc.cpp:551]:
```suggestion
                        Ctx, std::move(FOut), /*asmverbose=*/ true,
                        /*useDwarfDirectory=*/ true, IP, std::move(CE),
```

**Comment by arsenm** [comment]
> > > Utilize `-SAME:` directives?

If you use -SAME, you still have to produce 2 check lines for the 2 entirely different encoding comment formats. That's not really an improvement if the printed format isn't identical

File: llvm/lib/MC/MCHexStreamer.cpp (hunk 2 above)
Comment (kosarev) [reply llvm/lib/MC/MCHexStreamer.cpp]: Done.

File: llvm/lib/MC/MCHexStreamer.cpp (hunk 1 above)
Comment (kosarev) [reply llvm/lib/MC/MCHexStreamer.cpp]: Done.

File: llvm/lib/MC/MCHexStreamer.cpp (hunk 3 above)
Comment (kosarev) [reply llvm/lib/MC/MCHexStreamer.cpp]: Done.

File: llvm/tools/llvm-mc/llvm-mc.cpp (hunk 4 above)
Comment (kosarev) [reply llvm/tools/llvm-mc/llvm-mc.cpp]: Done.

**Review by kosarev — COMMENTED** [review]
> llvm-objdump seems to places the encodings as comments. You can check them with -SAME:.

Not until it's disassembled, I'm afraid, which is another RUN line, which is invovenient. And so is having two check lines where one would do, as Matt pointed out. Having two separate check lines also means they may match what is not a complete instruction, e.g., `flat_load_d16_b16 v5, v[1:2]` where the actual instruction may be `flat_load_d16_b16 v5, v[1:2] offset:0`.

These details apart, I also must admit I don't quite follow this view that testing is not something of sufficient importance to be properly reflected in our code infrastructure and that it should instead seek its ways around, such as by using `llvm-objdump`. `llvm-mc` is, by design, a tool for internal use, testing and debugging, our interface to the MC layer, and I would expect it to cover whatever use cases we developers have for it. Same for specifically the MCAsmStreamer's logic emitting instruction codes, which by the way includes verbalisation of relocations/fix-ups that I'm not sure how we could test using `llvm-objdump`.

**Comment by kosarev** [comment]
Ping.

**Comment by MaskRay** [comment]
> Not until it's disassembled, I'm afraid, which is another RUN line, which is invovenient. And so is having two check lines where one would do, as Matt pointed out. Having two separate check lines also means they may match what is not a complete instruction, e.g., flat_load_d16_b16 v5, v[1:2] where the actual instruction may be flat_load_d16_b16 v5, v[1:2] offset:0.
>
> These details apart, I also must admit I don't quite follow this view that testing is not something of sufficient importance to be properly reflected in our code infrastructure and that it should instead seek its ways around, such as by using llvm-objdump. llvm-mc is, by design, a tool for internal use, testing and debugging, our interface to the MC layer, and I would expect it to cover whatever use cases we developers have for it. Same for specifically the MCAsmStreamer's logic emitting instruction codes, which by the way includes verbalisation of relocations/fix-ups that I'm not sure how we could test using llvm-objdump.

Not sure I am following. It seems that `-filetype=hex` output is piped to `llvm-mc -disassemble`.

```
llvm-mc -filetype=hex a.s | llvm-mc -disassemble
```

How is this more convenient than `llvm-mc -filetype=obj a.s | llvm-objdump -d -`?

Can you analyze a mini example including the assembly source and llvm-mc/llvm-objdump command lines for clearer evaluation?

**Comment by kosarev** [comment]
> Not sure I am following. It seems that `-filetype=hex` output is piped to `llvm-mc -disassemble`.
> ```
> llvm-mc -filetype=hex a.s | llvm-mc -disassemble
> ```
> How is this more convenient than `llvm-mc -filetype=obj a.s | llvm-objdump -d -`?

I think this was already asked and answered above. To reiterate, `llvm-mc -disassemble` produces its output in the same format as `llvm-mc`, so we can use the same check directives for both assembler and dissasembler tests. Using `llvm-objdump -d -` would require multiple directives per instruction, which is not an improvement -- on top of other inconveniences explained earlier.

> Can you analyze a mini example including the assembly source and llvm-mc/llvm-objdump command lines for clearer evaluation?

I'm not sure I understand the question. llvm/test/MC/AMDGPU/gfx12_asm_vop1.s in this pull request should hopefully serve as an example, if that's the ask.

**Comment by MaskRay** [comment]
#92895, which does not add another MCStreamer, looks fine to me.
Note, we really should not add another MCStreamer, which is very easy to be misused.

MCAsmStreamer has some existing issues related to expression evaluation. If another MCStreamer is added, I am concerned others might use it for more assembler tasks (e.g. to support `.section`), which make future maintenance more difficult.
If the llvm-objdump output isn't convenient for AMDGPU, I think it is fine to add a new output mode.

**Comment by arsenm** [comment]
Can this be closed now?

**Comment by kosarev** [comment]
Abandoning this in favour of https://github.com/llvm/llvm-project/pull/92895.

**Pull request closed without merging**